raiden/connection_manager.py

```diff
@@ -292,17 +294,29 @@ def _open_channels(self):
         for partner in self.find_new_partners(qty_channels_to_open):
             try:
                 self.api.channel_open(
-                    self.token_network_identifier,
+                    self.registry_address,
+                    self.token_address,
                     partner,
                 )
             except DuplicatedChannelError:
                 # This can fail because of a race condition, where the channel
                 # partner opens first.
                 log.info('partner opened channel first')
 
+            exception = BaseException('Internal state did not sync an existing channel')
+            with gevent.Timeout(self.channel_sync_state_timeout, exception=exception):
+                waiting.wait_for_newchannel(
```

Comment:
[`RaidenApi.channel_open`](https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/api/python.py#L231-L238) is already calling `wait_for_newchannel`. This is not be necessary

Reply:
ha, this one was a tricky bug. But there's a possible sync problem:

if nodes A and B are running asynchronously, following can happen
1) A opens a channel
2) B also wants to open a channel at the same time, but A's transaction is already mined -> `self.api.channel_open()` raises `DuplicatedChannelError` (transaction throws)
https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/connection_manager.py#L298
3) BUT B's state isn't yet synced to a block in which the A's open channel transaction is mined
4) if B proceeds without waiting, and calls `self.api.set_total_channel_deposit()`, an exception will be raised from here
https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/api/python.py#L290
even if the channel exists on-chain (but not in the raiden state)

Reply:
OHHH, I see.

We should move the fix down to the `RaidenAPI` class then. Otherwise things like the rest api will also be susceptible to this race condition.

What about doing all of the following?

- Adding a check in the [`channel_open` method](https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/api/python.py#L187) to see if the channel already exists in the current state of the node, and if it does `DuplicatedChannelError` is raised
- Adding a [`try-except` around the proxy call](https://github.com/raiden-network/raiden/blob/c72407fd6c7d1f6e01dc703a1ba6278091038ca2/raiden/api/python.py#L222-L225), which handles the `DuplicatedChannelError` exception, calls `waiting.wait_for_newchannel` to make sure the node that lost the race is synchronized, and re-raises the exception

I think this will fix the race condition for the connection manager and the other users.

Reply:
@pcppcp Very nice find! Indeed we got a potential race there.

I second Augusto's suggestion to move the waiting logic/locking in the RaidenAPI so we cover all usage cases and not just this test.